### PR TITLE
[#19172][fix] Fix prompt embedding table H2D source lifetime

### LINKED ISSUE #19172: [Bug]: Async prompt-embedding H2D outlives its pooled-pinned source; the pinned pool can overwrite the buffer mid-copy
### System Info

- **TensorRT-LLM:** Both 1.1.0 and 1.2.1. Reproduced with the stock libtensorrt_llm.so shipped in the container, and with a locally built .so from the same tag.
- **GPU:** NVIDIA H100 80GB HBM3, sm90.
- **CUDA / driver:** 13.1
- **Container / backend:** Triton Inference Server, TensorRT-LLM C++ backend (inflight fused batching); classic TensorRT backend 26.02 and 26.07-trtllm-python-py3
- **Process topology (relevant):** the decoder runs in trtllmExecutorWorker, i.e. a separate process and therefore a separate CUDA context from the other Triton models and Python backend stubs on the same GPU.
- **Affected path:** request-supplied prompt embedding table (prompt tuning), non-offloaded, where the host source is held in the reusable pinned-memory pool (cpp/tensorrt_llm/runtime/tllmBuffers.h).

This is a **host-side object-lifetime defect, not a kernel or architecture issue.** It should be hardware-independent. High concurrency and a competing CUDA context on the same device are only needed to widen the window enough to observe it.

### Who can help?

Owners of:

cpp/tensorrt_llm/batch_manager/llmRequest.cpp — LlmRequest::movePromptEmbeddingTableToGpu()
cpp/tensorrt_llm/batch_manager/promptTuningBuffers.cpp — PromptTuningBuffers::fill()
cpp/tensorrt_llm/runtime/bufferManager.cpp — BufferManager::copyFrom() / copy()
cpp/tensorrt_llm/runtime/tllmBuffers.h — pinned pool

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

<p dir="ltr" style="">There are two ways to hit this. The first is deterministic and does not require the race to fire.</p>

<p dir="ltr" style=""><strong>A. Deterministic demonstration (recommended for review)</strong></p>

<p dir="ltr" style="">The regression test added in PR #18488 gates the CUDA stream with <code>cudaLaunchHostFunc</code> before the H2D, then asserts on source lifetime:</p>

<ol dir="ltr" style="">
<li>Insert a host callback that blocks the stream.</li>
<li>Create a pooled-pinned prompt embedding table; keep a <code>weak_ptr</code> to it.</li>
<li>Call <code>PromptTuningBuffers::fill()</code>, which queues the move and the event behind the gate.</li>
<li>Destroy or cancel the request.</li>
<li>On unpatched <code>main</code>, the <code>weak_ptr</code> has already expired — the source is back in the pool while the H2D is still pending on the stream.</li>
</ol>

<p dir="ltr" style="">This isolates the bug from scheduling luck entirely.</p>

<p dir="ltr" style=""><strong>B. Statistical reproduction (how it was originally found)</strong></p>

<ol dir="ltr" style="">
<li>Serve a model with prompt tuning enabled so that each request supplies its own prompt embedding table from pinned/pooled host memory (ours was 94,208 bytes per request).</li>
<li>Replay a fixed corpus of captured requests at concurrency 16 against the C++ backend with inflight batching. We used 3,108 requests per draw.</li>
<li>Add a <strong>competing CUDA process on the same GPU</strong>. It needs no access to TRT-LLM memory at all — eight Python backend instances running:</li>
</ol>

<div role="group" aria-label="python code" tabindex="0" style=""><div><div></div></div><div>python</div><div><pre style="color: rgb(234, 236, 240); background: transparent; font-family: var(--font-mono);"><code class="language-python" style="color: rgb(234, 236, 240); background: transparent; font-family: var(--font-mono); white-space: pre;"><span><span>x = torch<span style="color: rgb(211, 215, 222);">.</span>ones<span style="color: rgb(211, 215, 222);">(</span>N<span style="color: rgb(211, 215, 222);">,</span> <span style="color: rgb(247, 171, 95);">device</span>=<span style="color: rgb(155, 233, 99);">'cuda'</span><span style="color: rgb(211, 215, 222);">)</span>
</span><span>y = x + <span style="color: rgb(94, 237, 237);">1</span></span></span></code></pre></div></div>

<ol dir="ltr" start="4" style="">
<li>Classify outputs against a concurrency-1 golden capture.</li>
</ol>

<p dir="ltr" style="">Observed rates on our workload (each row 3,108 requests unless noted):</p>

<div dir="ltr" style="">
Configuration | Corrupted outputs
-- | --
Decoder only, C=16 | 0 / 3,108
Decoder only, C=48 | 0 / 3,108
+ CPU-only sidecar (same work, no CUDA) | 0 / 3,108
+ GPU sidecar process with an initialised CUDA context but no kernels | 0 / 3,108
+ dummy ones+1 CUDA sidecar, no synchronize | 10 / 3,108
+ dummy ones+1 CUDA sidecar with synchronize | 15 / 3,108
+ a real GPU neighbour model (×5 draws) | 252 / 15,540

</div>

<p dir="ltr" style="">Amplifier, if you need a higher yield for tracing: a 2 ms host sleep at the start of <code>decoderSync()</code> — after the launches, before the event wait — raised the rate to 28 / 3,108. A stall placed <em>inside</em> the stream (a fill-kernel <code>nanosleep</code>) does <strong>not</strong> amplify it, which is itself diagnostic: the vulnerable interval is between enqueue and the host's synchronisation point.</p>

<p dir="ltr" style="">Note that <code>CUDA_LAUNCH_BLOCKING=1</code> applied process-wide <strong>masks</strong> the bug (0 / 3,108, at ~24× the wall time), and so does MPS. Neither is a fix; both just close the window.</p>

### Expected behavior

BufferManager::copyFrom() issues cudaMemcpyAsync from a pinned host source. CUDA requires that source to remain allocated, unchanged, and at the same address until the copy completes on the stream.

The request-owned GPU prompt embedding table should therefore be a byte-exact copy of the host source, and greedy decoding over identical inputs should be deterministic regardless of what other processes are doing on the GPU.

### actual behavior

<p dir="ltr" style="">The GPU copy of the prompt embedding table is <strong>torn</strong>: it contains a mixture of the original request's bytes and a later request's bytes.</p>

<p dir="ltr" style="">Evidence from the run that localised it (one trapped request out of 3,108, greedy decoding, single repeated utterance so that every stage is deterministic and comparable to a C=1 golden):</p>

<div dir="ltr" style="">
Stage | Observation | Status
-- | -- | --
CPU source, after the Triton handoff | golden hash 0xf890b624ba27d041 | correct
Request-owned GPU table (first 16-byte canary) | 0x4465c937bd3dc1**00** vs golden 0x4465c937bd3dc1**ca** | torn
Shared GPU prompt buffer after the D2D | identical torn value | faithful copy of a bad input
Layer-0 KV after context | repeatable wrong hash 0x70837f2fce526439 vs golden 0xef0898cb6c2e1c4f | downstream
Engine logits | argmax is already the stop token | downstream
Sampled token | stop token, emitted after 1 generated token | user-visible failure

</div>

<p dir="ltr" style="">The Python/Triton input was correct and the host source was correct. The first divergence is the H2D itself, or the lifetime of its source during the H2D.</p>

<p dir="ltr" style="">Two independent causal checks:</p>

<ul dir="ltr" style="">
<li><strong>Retain the source</strong>: keeping the original pooled-pinned tensor alive on <code>LlmRequest</code> until the request ends — with <strong>no added stream synchronisation</strong> and no change to kernel math, sampler logic, or scheduling — removed both the torn canary and the output failure (0 / 3,108).</li>
<li><strong>Remove the copy</strong>: supplying the prompt embedding table already on the GPU (via DLPack from an upstream model), so <code>movePromptEmbeddingTableToGpu()</code> is never called for that input, also gave 0 / 3,108 — <strong>on the completely unmodified stock library</strong>, with the sidecar still running at ~4,100 inferences/s.</li>
</ul>

<p dir="ltr" style="">Together these bracket the defect to the pooled-host → request-GPU transfer boundary.</p>

<p dir="ltr" style=""><strong>Why this is easy to miss:</strong> the pinned block is returned to a pool, not unmapped, so the address stays valid. ASAN and Compute Sanitizer see a legal DMA from a mapped pinned region — no use-after-free, no illegal access. The result is corrupted-but-structurally-valid float data, so the symptom reads as model nondeterminism or sampling variance rather than memory corruption.</p>

### additional notes

Root cause — three individually reasonable contracts composing into a violation:

BufferManager::copyFrom() queues cudaMemcpyAsync on mStream and returns the destination. The caller must keep the source alive until the stream reaches the copy — but that obligation is implicit, and the method is named copyFrom, not copyFromAsync.
The pinned pool treats a released allocation as immediately reusable, and may hand the same address to the next request.
LlmRequest::movePromptEmbeddingTableToGpu() replaces mPromptEmbeddingTable with the GPU tensor as soon as copyFrom() returns, dropping the last host reference.

The third assumption violates the first whenever the source came from the second.

Failing sequence:

1  Request owns mPromptEmbeddingTable -> host pinned-pool tensor
2  BufferManager allocates a GPU destination
3  BufferManager queues cudaMemcpyAsync(dst, src) on its stream
4  copyFrom() returns to the CPU        <-- the GPU may not have read a byte yet
5  LlmRequest replaces the member with the GPU TensorPtr
6  The last host TensorPtr reference disappears
7  The pinned pool marks the segment free; the address becomes reusable
8  Another request receives that address and writes new bytes
9  The original H2D reads a mixture of old and new bytes
10 The GPU destination stays allocated, containing corrupted data

Steps 5–8 all occur inside the interval in which CUDA requires the source to be stable. Nothing in the C++ type system, the allocator, or CUDA itself represents that interval.

Repository archaeology (in case this looks older than it is): BufferManager::copyFrom is present in the initial commit (2023-09-20) and pinnedPool was added 2024-01-09, but the exact copy-then-replace-the-source-member shape in movePromptEmbeddingTableToGpu() dates from 2025-03-11. All three pieces are required, which is one reason this survived in a heavily used stack.

Why most deployments never hit it — six conditions must hold at once:

request-supplied prompt embedding tables in use at all;
the source arriving from the reusable pinned pool;
the last source owner dropped immediately after queueing the H2D;
pool reuse landing before the DMA finishes reading;
enough GPU scheduling pressure to stretch that interval;
the torn bytes changing the output detectably.

Proposed fix — PR #18488. PromptTuningBuffers owns a list of PendingPromptEmbeddingTableH2DCopy holders. Each owns the pooled-pinned source, a CudaEvent, and a flag recording whether completion was successfully recorded. Hot path: link the holder → queue the H2D → record the event on the BufferManager stream → continue without blocking → poll with cudaEventQuery on later fill() calls → release the source only when the event reports complete. The 1–3 ordering is deliberate so that no potentially-throwing ownership insertion remains after the copy is queued. Teardown waits on the exact copy events rather than the whole runtime stream. Where completion cannot be proven, the holder is quarantined (released without destroying the source) — a bounded leak on an exceptional path is preferable to returning possibly-in-flight pinned memory to the pool.

An earlier prototype that simply called cudaEventSynchronize() inline in movePromptEmbeddingTableToGpu() did fix the race on a targeted reproducer, but hung the full pipeline (GPU utilisation 0%, client timeouts) when integrated. The CUDA event was never the problem — the placement and consumption policy were. Request cancellation and destruction must stay nonblocking, and a request destructor must never block an executor thread, which is why the guard lives in PromptTuningBuffers rather than on LlmRequest.

Related audit candidates (not claims of defect): LlmRequest::moveLoraWeightsToGpu() has the identical copy-then-replace shape; whether it is exploitable depends on source memory type, pooling, and later stream ordering, and was not exercised by these runs. Other copyFrom call sites worth the same four questions — is the transfer async for this source/destination pair, who owns the source after the call returns, can that storage be mutated or pooled before the stream reaches the copy, and is completion guaranteed indirectly — include temporary vectors, request word lists, Medusa tables, encoder inputs, and executor tensor copies.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by rpengms**
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved reliability of asynchronous prompt-tuning data transfers by retaining source buffers until GPU copies complete.
  * Added safer handling for pending transfers during cleanup and shutdown.
  * Added CUDA event completion tracking for synchronization.

* **Tests**
  * Added coverage verifying prompt-embedding data remains available until asynchronous transfers finish.

[#19172][fix] Fix prompt embedding table H2D source lifetime
Co-authored-by: OpenAI Codex <codex@openai.com>

## Description

Fixes #19172

## Problem

`BufferManager::copyFrom` performs host-to-device copies asynchronously.
After moving a request's prompt embedding table to the GPU, `LlmRequest`
immediately dropped its reference to the pooled-pinned host tensor.

If the request was cancelled, destroyed, or otherwise released before the
H2D completed, the pinned pool could reuse and overwrite the source while
CUDA was still consuming it.

## Impact

Under high GPU concurrency, especially with a competing GPU sidecar,
scheduling delays enlarged this race window. The copied prompt embeddings
could become corrupted, producing nondeterministic output and random early
EOS even in greedy decoding.

The issue affects the non-offloaded prompt embedding table path when the
source is held in the reusable pinned-memory pool.

## Root cause

Three individually reasonable lifetime assumptions interacted incorrectly:

- `BufferManager` queues the copy asynchronously and requires the caller to
  keep the source alive.
- The pinned pool permits a released allocation to be reused immediately.
- `LlmRequest` replaced and released its host tensor without tracking H2D
  completion.

## Fix

- Retain the host tensor in `PromptTuningBuffers` before queueing the H2D.
- Record a CUDA event immediately after the copy.
- Poll events nonblockingly during subsequent buffer preparation.
- Release each source only after its event reports completion.
- During destruction, wait only on the exact copy events rather than the
  entire runtime stream.
- Quarantine the source on exceptional paths where completion cannot be
  proven.

Request cancellation and destruction remain nonblocking.

## Testing

Added a deterministic regression test that gates the CUDA stream and
verifies:

1. The request can release its prompt embedding table without releasing the
   H2D source.
2. An incomplete-event cleanup poll retains the source.
3. The source is released after the copy event completes.

Local CUDA tests were not run because this checkout has no configured C++
CUDA test build.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [ ] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by thorjohnsen — CHANGES_REQUESTED** [review]
Before you can start the CI pipeline, you need to make some procedural changes:

1. Sign off all commits. This is missing and is causing DCO test to fail, which will block CI pipeline.
2. Review the PR checklist and check the box.

After this, run CI pipeline "/bot run --disable-fail-fast" and fix all the issues that pop up. Let me know if you need help to determine if failures are caused by your code or flaky tests.

**Comment by rpengms** [comment]
> Before you can start the CI pipeline, you need to make some procedural changes:
> 
> 1. Sign off all commits. This is missing and is causing DCO test to fail, which will block CI pipeline.
> 2. Review the PR checklist and check the box.
> 
> After this, run CI pipeline "/bot run --disable-fail-fast" and fix all the issues that pop up. Let me know if you need help to determine if failures are caused by your code or flaky tests.

@thorjohnsen 

Thank you so much! I'm a little bit confused. - Is TRT-LLM team still accepting cpp fixes and will do official releases?

I thought cpp backend is dropped from 1.3 on, and will no longer be supported. That's why the PR target branch is a release branch **(release/1.2.1)** instead of main, cause in main the cpp code were deleted. (See this commit: [TRTLLM-14026][feat] BREAKING: Remove C++ modules for legacy TRT backend ([#16369](https://github.com/NVIDIA/TensorRT-LLM/pull/16369))) https://github.com/NVIDIA/TensorRT-LLM/commit/c77bc6ed6b044f48c174e786b5a5f7fafeee1728

That's why we're trying to just get a review of the proposed change content and plan to setup a private build on our side, to ship a private .so in our product.
But if you still release 1.1 and 1.2, that will be a different story, we can wait for official releases.
So, @thorjohnsen , please advice. We prefer to merge into v1.1, which target branch should this be in?

**Comment by thorjohnsen** [comment]
@rpengms I misunderstood your intent. We sometimes do patch releases initiated by customer PRs, but they are very rare and only done if they address critical security issues. We had an internal discussion, I don't think this PR qualifies as a critical security issue, partly because the code containing the bug has been deprecated in main. Your fix looks correct to me, but if you want a simpler way of doing this, have a look at https://github.com/NVIDIA/TensorRT-LLM/pull/15546. It fixes a similar issue by copying the input tensor to a temporary allocated with pytorch caching allocator. Pytorch manages the lifetime of any tensors allocated with caching allocator, it ensures all ops are done with it before it is reused. You don't have to change your implementation, it is already functionally correct.

**Comment by thorjohnsen** [comment]
@rpengms I misunderstood your intent. Your fix looks functionally correct to me, please go ahead and do a local build. Let me know if you need further assistance.

**Comment by rpengms** [comment]
> @rpengms I misunderstood your intent. Your fix looks functionally correct to me, please go ahead and do a local build. Let me know if you need further assistance.

Thanks a lot for doing the review and reply. We've successfully doing a private patch on our side.
